torch_xla/csrc/ir.h

```diff
@@ -0,0 +1,144 @@
+#pragma once
+
+#include <iostream>
+#include <memory>
+#include <set>
+#include <string>
+#include <utility>
+#include <unordered_map>
+#include <unordered_set>
+#include <vector>
+
+#include "tensorflow/compiler/xla/client/xla_builder.h"
+
+namespace torch_xla {
+namespace ir {
+
+class Node;
+class LoweringContext;
+
+using NodePtr = std::shared_ptr<Node>;
+
+// Represents a use of the output of a given node.
+// If use U is within node N, it means that node U.node is using the output
+// U.index of the node N.
+struct Use {
```

Comment:
I think memoizing the use information can wait until later. Fundamentally, it's just an optimization (we can always compute uses on query) and if we trim the `Node` interface as I've suggested there isn't an immediate use for it either.

Reply:
We need uses to walk back the graph, at the very least. Today to find parameters we need to lower to XLA, which is kinda ugly and slow.

Reply:
I agree w/r/t parameters, still not convinced we already need to go the extra mile to more efficiently track uses.

Reply:
Mile? It is like 15 yards :)
You cannot walk back w/out uses.
When we evaluate a node, we have the sinks only.

Reply:
I see, we don't have easy access to the graph "roots" to recover the use information. I guess this can stay as is then.